tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,504 @@ keywords:
+  \   /        \          /     \
+ -1-> 5->null  3->null   4->null 0->null
+   \             \        \   /
+    \           3->null     0-> 4->null
+     \              \      /  
+      \           0->3->4->null
+       \          /
+    -1->0->3->4->5->null    (sorted)
+```
+
+## Approach for top-down Merge sort of Linked list
+1. Keep recursively dividing the list until there is only one node in the linked list. 
+2. Sort each sublist and merge each sorted sublist in a new array.
+
+### Steps to apply merge sort to a Linked list
+
+**Note: Implementation is based on Merge Sort on an array as discussed above.** 
+1. Divide the linked list into two equal parts until when only one element is left.
+2. To Divide, we need to  find mid in the linked list using **slow and fast pointers** method.
+3. Then merge the left and the right nodes of the linked list.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+/*
+* Definition for singly-linked list.
+* struct ListNode {
+*     int val;
+*     ListNode *next;
+*     ListNode() : val(0), next(nullptr) {}
+*     ListNode(int x) : val(x), next(nullptr) {}
+*     ListNode(int x, ListNode *next) : val(x), next(next) {}
+* };
+*/
+class Solution {
+public:
+    void mergeSort(ListNode **head){
+        // create a current pointer
+        ListNode *curr = *head ; 
+        
+        ListNode * left;
+        ListNode * right;
+
+        // if the linked list is null 
+        // or the size is one return back the same
+        if(curr ==NULL || curr ->next==NULL) return; 
+
+        //call to find the middle node between left and right
+        findMiddle(curr, &left,&right);
+        
+        // call merge_sort again to divide left half
+        mergeSort(&left);
+        
+        // call merge_sort again to divide right half
+        mergeSort(&right);
+        
+        //call to merge left and right by sorting them
+        *head=merge(left,right);
+    }
+
+    void findMiddle(ListNode *curr,ListNode **left, ListNode **right){
+        // make a slow pointer
+        ListNode* slow = curr;
+
+        // make a fast pointer 
+        ListNode* fast = curr -> next; 
+        
+    // then we move our fast up to it not become null
+        while(fast != NULL)
+        {
+            fast = fast -> next;
+            if(fast != NULL)
+            {
+                fast = fast -> next;
+                slow = slow -> next;
+            }
+        }
+        *left = curr;
+        // right to slow next
+        *right = slow -> next;
+
+        // and slow next to null 
+        slow -> next = NULL;         
+    }
+
+    ListNode* merge(ListNode* left, ListNode* right){
+      
+        ListNode* res = NULL; 
+        // Check if left is null,nothing to merge
+        if(left == NULL) return right;
+        
+        if(right == NULL) return left;
+    
+        // if value of the left <= value of right
+        //then res = left
+        if(left -> val <= right -> val) 
```

Comment:
don't add space for arrow operator. please change all.

`left->val` and `right->val`